addons/project/models/project_project.py

```diff
@@ -463,15 +473,18 @@ def create(self, vals_list):
     def write(self, vals):
         # Here we modify the project's stage according to the selected company (selecting the first
         # stage in sequence that is linked to the company).
-        if self.env.user.has_group('project.group_project_stages') and\
-            vals.get('company_id') not in (None, self.company_id.id) and\
-            self.stage_id.company_id:
-            ProjectStage = self.env['project.project.stage']
-            vals['stage_id'] = ProjectStage.search(
-                [('company_id', 'in', (vals['company_id'], False))],
-                order=f"company_id asc, {ProjectStage._order}",
-                limit=1,
-            ).id
+        if self.env.user.has_group('project.group_project_stages') and vals.get('company_id'):
+            projects_already_with_company = self.filtered(lambda p: p.company_id.id == vals.get('company_id'))
+            if projects_already_with_company:
+                projects_already_with_company.write({key: value for key, value in vals.items() if key != 'company_id'})
+                self -= projects_already_with_company
+            if vals.get('company_id') not in (None, self.company_id.ids) and self.stage_id.company_id:
```

Comment:
```suggestion
            if vals.get('company_id') not in (None, self.company_id.id) and self.stage_id.company_id:
```
?

Reply:
I think it was the main problem causing the traceback. Since it can happen that there are projects with differents companies among self, I think we need to use `ids` : Here what's happening when calling `id` : 
`self.env["project.project"].search([]).company_id.id`

`Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/lucas/Documents/odoo/odoo/odoo/fields.py", line 5079, in __get__
    raise ValueError("Expected singleton: %s" % record)
ValueError: Expected singleton: res.company(1, 26)`

`self.env["project.project"].search([]).company_id.ids`
`[1, 26]`

Reply:
Yeah, makes more sense :)